smithy-model/src/main/java/software/amazon/smithy/model/loader/IdlModelParser.java

```diff
@@ -570,37 +584,120 @@ private void parseStructuredShape(
         ws();
 
         // Parse optional "with" statements to add mixins, but only if it's supported by the version.
-        if (peek() == 'w') {
-            expect('w');
-            expect('i');
-            expect('t');
-            expect('h');
+        parseMixins(id);
+        parseMembers(id, Collections.emptySet(), structureMember);
+    }
 
-            if (!modelFile.getVersion().supportsMixins()) {
-                throw syntax(id, "Mixins can only be used with Smithy version 2 or later. "
-                                 + "Attempted to use mixins with version `" + modelFile.getVersion() + "`.");
-            }
+    private void parseMixins(ShapeId id) {
+        if (peek() != 'w') {
+            return;
+        }
 
-            ws();
-            do {
-                String target = ParserUtils.parseShapeId(this);
-                modelFile.addForwardReference(target, resolved -> modelFile.addPendingMixin(id, resolved));
-                ws();
-            } while (peek() != '{');
+        expect('w');
+        expect('i');
+        expect('t');
+        expect('h');
+
+        if (!modelFile.getVersion().supportsMixins()) {
+            throw syntax(id, "Mixins can only be used with Smithy version 2 or later. "
+                    + "Attempted to use mixins with version `" + modelFile.getVersion() + "`.");
         }
 
-        parseMembers(id, Collections.emptySet(), structureMember);
+        ws();
+        do {
+            String target = ParserUtils.parseShapeId(this);
+            modelFile.addForwardReference(target, resolved -> modelFile.addPendingMixin(id, resolved));
+            ws();
+        } while (peek() != '{');
     }
 
     private void parseOperationStatement(ShapeId id, SourceLocation location) {
-        ws();
         OperationShape.Builder builder = OperationShape.builder().id(id).source(location);
-        ObjectNode node = IdlNodeParser.parseObjectNode(this, id.toString());
-        LoaderUtils.checkForAdditionalProperties(node, id, OPERATION_PROPERTY_NAMES, modelFile.events());
+        parseProperties(id, OPERATION_PROPERTY_NAMES, propertyName -> {
+            switch (propertyName) {
+                case "input":
+                    parseInlineableOperationMember(id, operationInputSuffix, builder::input);
+                    break;
+                case "output":
+                    parseInlineableOperationMember(id, operationOutputSuffix, builder::output);
+                    break;
+                case "errors":
+                    parseIdList(builder::addError);
+                    break;
+                default:
```

Comment:
Could we ditch the OPERATION_PROPERTY_NAMES stuff and the validation in parseProperties, and instead handle that here in the switch? It seems redundant.